src/agent/internals/src/main/java/com/google/devtools/cdbg/debuglets/java/FirebaseClient.java

```diff
@@ -991,45 +1094,42 @@ private static Object getDbValue(
 
     infofmt("Beginning Firebase Database read operation at '%s'", path);
 
-    ValueEventListener listener = null;
-
     final ArrayBlockingQueue<Boolean> resultObtained = new ArrayBlockingQueue<>(1);
     final ArrayList<Object> obtainedValue = new ArrayList<>();
 
-    try {
-      dbRef.addValueEventListener(
-          new ValueEventListener() {
-            @Override
-            public void onDataChange(DataSnapshot dataSnapshot) {
-              Object value = dataSnapshot.getValue();
-              infofmt(
-                  "Response obtained, data was %s found at %s", value == null ? "not" : "", path);
-
-              obtainedValue.add(value);
-              resultObtained.offer(Boolean.TRUE);
-            }
+    dbRef.addListenerForSingleValueEvent(
+        new ValueEventListener() {
+          @Override
+          public void onDataChange(DataSnapshot dataSnapshot) {
+            Object value = dataSnapshot.getValue();
+            infofmt("Response obtained, data was %s found at %s", value == null ? "not" : "", path);
 
-            @Override
-            public void onCancelled(DatabaseError error) {
-              warnfmt("Database subscription error: %s", error.getMessage());
-              resultObtained.offer(Boolean.FALSE);
-            }
-          });
+            obtainedValue.add(value);
+            resultObtained.offer(Boolean.TRUE);
+          }
 
-      // Returns null on timeout.
-      Boolean isSuccess = resultObtained.poll(timeout.value, timeout.units);
+          @Override
+          public void onCancelled(DatabaseError error) {
+            warnfmt("Database subscription error: %s", error.getMessage());
+            resultObtained.offer(Boolean.FALSE);
+          }
+        });
 
-      // null will be returned on read timeout.
-      if (isSuccess == null) {
-        infofmt("Read from %s timed out after %d %s", path, timeout.value, timeout.units);
-        throw new Exception("Read timed out");
-      } else if (!isSuccess) {
-        throw new Exception("Error occurred attempting to read from the DB");
-      }
-    } finally {
-      if (listener != null) {
-        dbRef.removeEventListener(listener);
-      }
+    // Returns null on timeout.
+    Boolean isSuccess = resultObtained.poll(timeout.value, timeout.units);
+
+    // null will be returned on read timeout.
+    if (isSuccess == null) {
+      infofmt("Read from %s timed out after %d %s", path, timeout.value, timeout.units);
+      throw new Exception(
+          "Firebase Database read operation from '"
+              + path
+              + "' timed out after "
+              + timeout.value
+              + " "
+              + timeout.units.toString());
```

Comment:
Or just add a toString() to the Timeout class.

Reply:
good call